File: Sean的leetcode/python/Implement Trie (Prefix Tree).py
###
# Trie
# Time Complexity: insert - O(m) search - O(m) startswith - O(m) m: len of key
# Space Complexity: insert - O(m) search - O(1) startswith - O(1)
###
class TrieNode(object):
    def __init__(self):
        self.children = collections.defaultdict(TrieNode)
        self.is_end = False
        
class Trie(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()

    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: void
        """
        cur = self.root
        for c in word:
            cur = cur.children[c]
        cur.is_word = True
        
    def find(self, word):
        cur = self.root
        for c in word:
            if c not in cur.children:
                return None
            cur = cur.children[c]
        return cur
    
    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """
        node = self.find(word) 
        return node and node.is_word
        

    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        node = self.find(prefix)
        return node is not None


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
########
# If only en characters, we can also use an array instead of a dict
########
class TrieNode(object):
    def __init__(self):
        self.is_end = False
        self.children = [None] * 26


class Trie(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()

    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: None
        """
        cur = self.root
        for c in word:
            idx = ord(c) - ord('a')
            if not cur.children[idx]:
                cur.children[idx] = TrieNode()
            cur = cur.children[idx]
        cur.is_end = True

    def find(self, word):
        cur = self.root
        for c in word:
            idx = ord(c) - ord('a')
            if not cur.children[idx]:
                return None
            cur = cur.children[idx]
        return cur

    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """
        cur = self.find(word)
        return cur and cur.is_end

    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        cur = self.find(prefix)
        return cur is not None

# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: MyLeetCode/Leetcode/Implement Trie (Prefix Tree).cpp
class TrieNode {
public:
    // Initialize your data structure here.
    TrieNode() : isWord(false) {
        child = vector<TrieNode*>(26);
        
    }
    bool isWord;
    vector<TrieNode*> child;
};

class Trie {
public:
    Trie() {
        root = new TrieNode();
    }

    // Inserts a word into the trie.
    void insert(string word) {
        TrieNode* curr = root;
        for (auto c : word) {
            if (!curr->child[c - 'a']) {
                curr->child[c - 'a'] = new TrieNode();
            }
            curr = curr->child[c - 'a'];
        }
        curr->isWord = true;
    }

    // Returns if the word is in the trie.
    bool search(string word) {
        return find(word, true);
        
    }

    // Returns if there is any word in the trie
    // that starts with the given prefix.
    bool startsWith(string prefix) {
        return find(prefix, false);
    }

private:
    TrieNode* root;
    bool find(string word, bool end) {
        TrieNode* curr = root;
        for (auto c : word) {
            if (!curr->child[c-'a']) {
                return false;
            }
            curr = curr->child[c - 'a'];
        }
        return end? curr->isWord : true;
    }
};

// Your Trie object will be instantiated and called as such:
// Trie trie;
// trie.insert("somestring");
// trie.search("key");
//Sean's solution
class TrieNode {
public:
    unordered_map<char, TrieNode*>dict;
    int end;

    // Initialize your data structure here.
    TrieNode() {
        end = 0;
    }
};

class Trie {
public:
    Trie() {
        root = new TrieNode();
    }

    // Inserts a word into the trie.
    void insert(string word) {
        TrieNode *curr = root;
        for (int i = 0; i < word.size(); i++)
        {
            if (!curr->dict.count(word[i]))
                curr->dict[word[i]] = new TrieNode();
            curr = curr->dict[word[i]];
        }
        curr->end = 1;
    }

    // Returns if the word is in the trie.
    bool search(string word) {
        return find(word, 1);
    }

    // Returns if there is any word in the trie
    // that starts with the given prefix.
    bool startsWith(string prefix) {
        return find(prefix, 0);
    }

private:
    TrieNode* root;
    bool find(string s, int is_end)
    {
        TrieNode *curr = root;
        for (int i = 0; i < s.size(); i++)
        {
            if (!curr->dict.count(s[i]))
                return false;
            curr = curr->dict[s[i]];
        }
        return is_end ? curr->end == 1 : true;
    }
};

// Your Trie object will be instantiated and called as such:
// Trie trie;
// trie.insert("somestring");
// trie.search("key");